website/payments/api/v2/views.py

```diff
@@ -0,0 +1,145 @@
+        "GET": ["payments:read"],
+        "POST": ["payments:write"],
+    }
+    filter_backends = (
+        framework_filters.OrderingFilter,
+        filters.CreatedAtFilter,
+        filters.PaymentTypeFilter,
+    )
+    ordering_fields = ("created_at",)
+
+    def get_queryset(self):
+        return (
+            super()
+            .get_queryset()
+            .filter(paid_by=PaymentUser.objects.get(pk=self.request.member.pk))
+        )
+
+
+class PaymentDetailView(RetrieveAPIView):
+    queryset = Payment.objects.all()
+    serializer_class = PaymentSerializer
+    permission_classes = [IsAuthenticatedOrTokenHasScopeForMethod]
+    required_scopes_per_method = {"GET": ["payments:read"]}
+
+    def get_queryset(self):
+        return (
+            super()
+            .get_queryset()
+            .filter(paid_by=PaymentUser.objects.get(pk=self.request.member.pk))
+        )
+
+
+class PayableDetailView(RetrieveAPIView):
+    permission_classes = [IsAuthenticatedOrTokenHasScopeForMethod]
+    required_scopes_per_method = {
+        "GET": ["payments:read"],
+        "PUT": ["payments:write"],
+        "PATCH": ["payments:write"],
+    }
+
+    def get_serializer_class(self, *args, **kwargs):
+        if self.request.method.lower() == "put":
+            return EmptySerializer
+        return PayableSerializer
+
+    def get(self, request, *args, **kwargs):
+        serializer = self.get_serializer(self.get_payable())
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def get_payable(self):
+        app_label = self.kwargs["app_label"]
+        model_name = self.kwargs["model_name"]
+        payable_pk = self.kwargs["payable_pk"]
+
+        try:
+            payable_model = apps.get_model(app_label=app_label, model_name=model_name)
+            payable = payables.get_payable(
+                get_object_or_404(payable_model, pk=payable_pk)
+            )
+        except (LookupError, NotRegistered) as e:
+            raise serializers.ValidationError(
+                {api_settings.NON_FIELD_ERRORS_KEY: [_("Payable model not found")]}
+            ) from e
+
+        if (
+            not payable.payment_payer
+            or not payable.tpay_allowed
+            or payable.payment_payer != PaymentUser.objects.get(pk=self.request.user.pk)
+        ):
+            raise PermissionDenied(
+                detail=_("You do not have permission to perform this action.")
+            )
+
+        return payable
+
+    def delete(self, request, *args, **kwargs):
+        payable = self.get_payable()
+
+        if not payable.model.payment:
+            raise Http404
+
+        try:
+            services.delete_payment(
+                payable.model, request.member,
+            )
+            payable.model.save()
+        except PaymentError as e:
+            raise PermissionDenied(detail=str(e))
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def put(self, request, *args, **kwargs):
```

Comment:
Payments are immutable so I don't think there should be a PUT, just a POST

Reply:
good point :D